File: algorithms/sorting/quick.py
"""
=====================================================================
Quick Sort / Быстрая сортировка
---------------------------------------------------------------------
Идея (divide & conquer):
  1. Выбираем pivot
  2. Partition: < pivot влево, >= вправо
  3. Рекурсивно сортируем части
Time:  O(n log n) avg, O(n²) worst
Space: O(log n) avg (стек рекурсии)
Stable: no
In-place: yes
=====================================================================
"""
from typing import List


def quick_sort(arr: List[int], lo: int = 0, hi: int = None) -> List[int]:
    if hi is None:
        hi = len(arr) - 1

    if lo < hi:
        p = partition(arr, lo, hi)
        quick_sort(arr, lo, p - 1)
        quick_sort(arr, p + 1, hi)

    return arr


def partition(arr: List[int], lo: int, hi: int) -> int:
    """Lomuto partition: arr[hi] — pivot."""
    pivot = arr[hi]
    i = lo - 1

    for j in range(lo, hi):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i + 1], arr[hi] = arr[hi], arr[i + 1]
    return i + 1


if __name__ == "__main__":
    print(quick_sort([10, 7, 8, 9, 1, 5]))  # [1, 5, 7, 8, 9, 10]

File: algorithms/sorting/quick.js
// =====================================================================
// Quick Sort / Быстрая сортировка
// ---------------------------------------------------------------------
// Идея (divide & conquer):
//   1. Выбираем опорный (pivot)
//   2. Partition: элементы < pivot влево, >= вправо
//   3. Рекурсивно сортируем обе части
// Time:  O(n log n) avg, O(n²) worst (плохой pivot)
// Space: O(log n) avg на стек рекурсии
// Stable: no
// In-place: yes (Lomuto partition ниже)
// Совет: выбирай pivot случайно или медиану из трёх, чтобы избежать O(n²)
//        на отсортированных данных.
// =====================================================================

function quickSort(arr, lo = 0, hi = arr.length - 1) {
    if (lo < hi) {
        // p — финальная позиция pivot после partition
        const p = partition(arr, lo, hi);

        // Рекурсивно сортируем левую и правую части (без pivot)
        quickSort(arr, lo, p - 1);
        quickSort(arr, p + 1, hi);
    }
    return arr;
}

/**
 * Lomuto partition scheme.
 * Кладёт arr[hi] (pivot) на правильное место, возвращает его индекс.
 */
function partition(arr, lo, hi) {
    const pivot = arr[hi];
    let i = lo - 1; // граница "меньше pivot"

    for (let j = lo; j < hi; j++) {
        if (arr[j] <= pivot) {
            i++;
            [arr[i], arr[j]] = [arr[j], arr[i]];
        }
    }

    // Ставим pivot на свою позицию
    [arr[i + 1], arr[hi]] = [arr[hi], arr[i + 1]];
    return i + 1;
}

console.log(quickSort([10, 7, 8, 9, 1, 5])); // [1, 5, 7, 8, 9, 10]

module.exports = { quickSort, partition };
